# project name
PROJECT_NAME: "lio_lite"

lio_lite:

  # Topics
  pointCloudTopic: "/livox/lidar"                # Point cloud data
  imuTopic: "/livox/imu"                         # IMU data
  odomTopic: "odometry/imu"                      # IMU pre-preintegration odometry, same frequency as IMU
  gpsTopic: "/gps/raw_data"                      # GPS odometry topic from navsat, see module_navsat.launch file
  
  # Frames
  lidarFrame: "base_link"
  baselinkFrame: "base_link"
  odometryFrame: "odom"
  mapFrame: "map"

  # Export settings
  savePCD: true                              
  savePCDDirectory: "temp"                    

  # Sensor Settings
  sensor: livox                               # lidar sensor type, 'velodyne' or 'ouster' or 'livox'  NNS:没啥作用
  N_SCAN: 4                                  # number of lidar channel (i.e., Velodyne/Ouster: 16, 32, 64, 128, Livox Horizon: 6)
  Horizon_SCAN: 6000                          # lidar horizontal resolution (Velodyne:1800, Ouster:512,1024,2048, Livox Horizon: 4000)
  timeField: "time"                           # point timestamp field, Velodyne - "time", Ouster - "t"
  downsampleRate: 1                           # default: 1. Downsample your data if too many points. i.e., 16 = 64 / 4, 16 = 16 / 1
  lidarMinRange: 0.5                          # default: 1.0, minimum lidar range to be used
  lidarMaxRange: 40.0                        # default: 1000.0, maximum lidar range to be used

  # IMU Settings
  imuAccNoise: 1.4262640130500099e-02
  imuGyrNoise: 3.7933965568774236e-03
  imuAccBiasN: 1.5343404030301728e-04
  imuGyrBiasN: 5.2031846544925295e-05
  imuGravity: 9.7940
  imuRPYWeight: 0.01
  
  initializedMode: 1                           # 重定位初始化模式：1为手动提供初值，2为自动初始化
  globalRelocalizeInitialFitnessScore: 0.25    # 重定位初始化的阈值：设置大一些对初值的要求会降低
  globalRelocalizeFitnessScore: 0.1            # 重定位运行时的阈值：设置小一些可以提高精度

  ###################### extrinsic between IMU and LiDAR  ###########################
  ###################### T_IMU_LiDAR, LiDAR -> IMU       ############################
  extrinsicTranslation: [0.0, 0.0, 0.0]    
  extrinsicRotation: [1,   0,  0, 
                      0,   1,  0, 
                      0,   0,  1]

  ########################## IMU property config  ###########################
  ## 对绝大多数IMU来说，下面三个值分别是"+z", "+y", "+x" (for most of IMUs, the following config is "+z", "+y", "+x")
  # 绕着哪个轴逆时针转动，输出yaw角度为正(which axis the IMU rotates around counterclockwise to get a positive yaw angle)
  yawAxis: "+z"  
  # 绕着哪个轴逆时针转动，输出pitch角度为正(which axis the IMU rotates around counterclockwise to get a positive pitch angle)
  pitchAxis: "+x"    
  # 绕着哪个轴逆时针转动，输出roll角度为正(which axis the IMU rotates around counterclockwise to get a positive roll angle)
  rollAxis: "+y"


  # LOAM feature threshold
  edgeThreshold: 0.3
  surfThreshold: 0.1
  edgeFeatureMinValidNum: 10
  surfFeatureMinValidNum: 100

  # voxel filter paprams
  odometrySurfLeafSize: 0.4                     # default: 0.4
  mappingCornerLeafSize: 0.2                    # default: 0.2
  mappingSurfLeafSize: 0.4                      # default: 0.4

  # robot motion constraint (in case you are using a 2D robot)
  z_tollerance: 1000                            # meters
  rotation_tollerance: 1000                     # radians

  # CPU Params
  numberOfCores: 12                             # number of cores for mapping optimization
  mappingProcessInterval: 0.15                  # seconds, regulate mapping frequency

  # Surrounding map
  surroundingkeyframeAddingDistThreshold: 1.0   # meters, regulate keyframe adding threshold
  surroundingkeyframeAddingAngleThreshold: 0.2  # radians, regulate keyframe adding threshold
  surroundingKeyframeDensity: 2.0               # meters, downsample surrounding keyframe poses   
  surroundingKeyframeSearchRadius: 50.0         # meters, within n meters scan-to-map optimization (when loop closure disabled)

  # Loop closure
  loopClosureEnableFlag: true
  loopClosureFrequency: 1.0
  surroundingKeyframeSize: 25                   # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 20.0             # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 60.0           # seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 25                  # number of hostory key frames will be fused into a submap for loop closure
  historyKeyframeFitnessScore: 0.1              # icp threshold, the smaller the better alignment

  # Visualization
  globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius
  globalMapVisualizationPoseDensity: 10.0       # meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 1.0           # meters, global map visualization cloud density